### PR TITLE
Implement UniqueSequences().

### PR DESCRIPTION
**Pull request by danpovey**
.. will be needed in n-best list processing.
Also extend `Index(RaggedShape, Array1..)` to take an `axis` argument.

### CODE REVIEW

File: k2/csrc/ragged_ops.h [hunk 1]
```diff
@@ -427,6 +412,14 @@ Ragged<int32_t> AddPrefixToRagged(Ragged<int32_t> &src,
  */
```
Comment (csukuangfj) [inline k2/csrc/ragged_ops.h:412]: > Read algorithms like in compose.cc to understand why.

What do you mean by saying `compose.cc`? There is no such file in k2.

File: k2/csrc/ragged_ops.h [hunk 2]
```diff
@@ -1321,6 +1331,25 @@ template <typename T>
...
+  the earliest axes will be ignored.
+
+     @param [in] src  Source ragged tensor
+
+     @return   Returns a tensor with the same number of axes as `src` and
+            possibly fewer elements due to removing repeated sequences on the
+            last axis (and with the last-but-one indexes possibly in a different
+            order).
+
+  CAUTION: the current implementation does not completely remove the possibility
```
Comment (csukuangfj) [inline k2/csrc/ragged_ops.h:1347]: If several sequences have the same hash, the current implementation
keeps only one of them.

Repeated sequences would have an identical hash value. Why does it say
> the current implementation does not completely remove the possibility of repeated sequnces,

File: k2/csrc/ragged_ops.h (hunk 1 above)
Comment (danpovey) [reply k2/csrc/ragged_ops.h:412]: can remove this sentence, must have moved things around.

File: k2/csrc/ragged_ops.h (hunk 2 above)
Comment (danpovey) [reply k2/csrc/ragged_ops.h:1347]: mm that was a mistake, it does not completely guarantee that all unique sequences will 
be present in the output.  Perhaps you can fix..

Comment (csukuangfj) [reply k2/csrc/ragged_ops.h:1347]: Will fix it.

File: k2/python/csrc/torch/ragged.cu [hunk 3]
```diff
@@ -151,7 +151,7 @@ static void PybindRaggedTpl(py::module &m, const char *name) {
         Array1<int32_t> indexes_array = FromTensor<int32_t>(indexes);
         Array1<int32_t> value_indexes;
 
-        Ragged<T> ans = Index(src, indexes_array,
+        Ragged<T> ans = Index(src, 0, indexes_array,
```
Comment (csukuangfj) [inline k2/python/csrc/torch/ragged.cu:154]: It does not handle the case when `indexes_array` contains -1, which will result in OOM.

Comment (danpovey) [reply k2/python/csrc/torch/ragged.cu:154]: Oh that's interesting.  I was just looking into the same issue for a different reason.

Comment (danpovey) [reply k2/python/csrc/torch/ragged.cu:154]: Let me fix it.

Comment (csukuangfj) [reply k2/python/csrc/torch/ragged.cu:154]: `index_array` is from the arc map, which may contain `-1`, violating the pre-condition of `Index`.

Comment (csukuangfj) [reply k2/python/csrc/torch/ragged.cu:154]: thanks

**Pull request merged**